### DISCUSSION TITLE
Is it possible to modulate the initial MonochromaticField?

### DISCUSSION THREAD

**Discussion by Windaway** [rafael-fuente/Diffraction-Simulations--Angular-Spectrum-Method#15] [Q&A]
Thank you for creating such a useful tool. 
In my recent research, I need a plane wave  only whose intensity is modulated, but I can't achieve this effect with MonochromaticField and Could you tell me how to implement this function?

**Comment by rafael-fuente** [comment]
Hi Windaway,

Do you need the modulated intensity in time or space? If it's in space it can be directly done with MonochromaticField.

If it's in time, we need to Fourier transform your signal and simulate each wavelength separately. I can build a diffractsim,  
interface for doing it since it's an interesting feature. Is that what you need?

**Comment by Windaway** [comment]
Thank you for your reply. In fact, I need the spatially modulated plane wave.  Before, I add a ApertureFromImage with the amplitude_mask_path, but that's not what I want. I read the code again and will try to change the MonochromaticField.E.

Thank you again.

**Comment by rafael-fuente** [comment]
Hi Windaway,

I'm curious to take a look at your code to check if everything is right, if you want.
Could you share your script?

**Comment by Windaway** [comment]
> Hi Windaway,
> 
> I'm curious to take a look at your code to check if everything is right, if you want. Could you share your script?

Here is the code.
```
import diffractsim
import cv2
import numpy as np
diffractsim.set_backend("CPU") #Change the string to "CUDA" to use GPU acceleration
from diffractsim import MonochromaticField, ApertureFromImage, Lens, mm, nm, cm, FourierPhaseRetrieval
F = MonochromaticField(
    wavelength=632.8 * nm, extent_x=30 * mm, extent_y=30 * mm, Nx=2400, Ny=2400, intensity = 0.005
)
img=cv2.resize (cv2.imread('mask.png',cv2.IMREAD_GRAYSCALE),(2400,2400),interpolation=cv2.INTER_CUBIC)
img=np.array(img,dtype=np.float32)/255.
for x in range(2400):
    for y in range(2400):
        F.E[x][y]=F.E[x][y]*img[x,y]
rgb = F.get_colors()
F.plot_colors(rgb)
F.add(ApertureFromImage(
    amplitude_mask_path= "ap.jpg", image_size=(30.0 * mm, 30.0 * mm), simulation = F))
rgb = F.get_colors()
F.plot_colors(rgb)
F.propagate(10*cm)
rgb = F.get_colors()
F.plot_colors(rgb)
```

**Comment by rafael-fuente** [comment]
> > Hi Windaway,
> > I'm curious to take a look at your code to check if everything is right, if you want. Could you share your script?
> 
> Here is the code.
> 
> ```
> import diffractsim
> import cv2
> import numpy as np
> diffractsim.set_backend("CPU") #Change the string to "CUDA" to use GPU acceleration
> from diffractsim import MonochromaticField, ApertureFromImage, Lens, mm, nm, cm, FourierPhaseRetrieval
> F = MonochromaticField(
>     wavelength=632.8 * nm, extent_x=30 * mm, extent_y=30 * mm, Nx=2400, Ny=2400, intensity = 0.005
> )
> img=cv2.resize (cv2.imread('mask.png',cv2.IMREAD_GRAYSCALE),(2400,2400),interpolation=cv2.INTER_CUBIC)
> img=np.array(img,dtype=np.float32)/255.
> for x in range(2400):
>     for y in range(2400):
>         F.E[x][y]=F.E[x][y]*img[x,y]
> rgb = F.get_colors()
> F.plot_colors(rgb)
> F.add(ApertureFromImage(
>     amplitude_mask_path= "ap.jpg", image_size=(30.0 * mm, 30.0 * mm), simulation = F))
> rgb = F.get_colors()
> F.plot_colors(rgb)
> F.propagate(10*cm)
> rgb = F.get_colors()
> F.plot_colors(rgb)
> ```

It's everything fine. But note that you don't really need to load 'mask.png' manually. It can be done way faster by simply using another `ApertureFromImage` object:



```python

import diffractsim
import numpy as np
diffractsim.set_backend("CPU") #Change the string to "CUDA" to use GPU acceleration
from diffractsim import MonochromaticField, ApertureFromImage, Lens, mm, nm, cm

F = MonochromaticField(
    wavelength=632.8 * nm, extent_x=30 * mm, extent_y=30 * mm, Nx=2400, Ny=2400, intensity = 0.005
)

F.add(ApertureFromImage(
    amplitude_mask_path= "mask.png", image_size=(30.0 * mm, 30.0 * mm), simulation = F))

rgb = F.get_colors()
F.plot_colors(rgb)

F.add(ApertureFromImage(
    amplitude_mask_path= "ap.jpg", image_size=(30.0 * mm, 30.0 * mm), simulation = F))

rgb = F.get_colors()
F.plot_colors(rgb)

F.propagate(10*cm)
rgb = F.get_colors()
F.plot_colors(rgb)
```
